### DISCUSSION TITLE
Convert .eslintrc.json to eslint.config.js

### DISCUSSION THREAD

**Discussion by fudom** [eslint/eslint#18543] [Ideas]
Just updated from 8 to 9 and fail on the config file. Why not just auto-convert it? Is there a converter?
I know the https://eslint.org/docs/latest/use/configure/migration-guide but auto convert would be better.
I wish to lint --fix .eslintrc.json to eslint.config.js

**Comment by archasek** [comment]
I'll just leave it here... simplify your life :) https://github.com/eslint/eslint/discussions/18207#discussioncomment-9538934

**Comment by mdjermanovic** [comment]
> Is there a converter?

Yes: https://www.npmjs.com/package/@eslint/migrate-config

**Reply by MrSuddenJoy** [reply]
I know Im not the one who asked for it, but was facing similar thing to what @fudom was facing. Many thanks @mdjermanovic for the link. Appreciate.

**Reply by fudom** [reply]
Thanks, but it's still impossible to get this to work. Maybe it only works on simple projects. My current solution or way is: 

```
npm init @eslint/config@latest
```

And rebuild the configuration step by step. Maybe leave out a few things, etc. I'm not sure if this approach will get me to my goal. But I think it's the better way than compat with never ending problems.

Btw, just out of interest: Why this huge breaking change? It seems more complicated now. But ok, I'm not yet done. There are good reasons for sure. But please understand that this causes problems and annoyances for users.

Quick side question: What exactly does each array entry of export default represent? I see a split configuration, etc. At first I thought these were something like overrides and defining linter configurations per file.

Update: ESLint does not work in VSCode anymore (no error). With CLI I see errors. But not visible in editor. Any guesses?

**Reply by mdjermanovic** [reply]
> Btw, just out of interest: Why this huge breaking change?

For reasons for introducing the new config system, please see:

* https://eslint.org/blog/2022/08/new-config-system-part-1/
* https://eslint.org/blog/2022/08/new-config-system-part-2/
* https://github.com/eslint/rfcs/tree/main/designs/2019-config-simplification
* https://github.com/eslint/eslint/issues/3458

> Quick side question: What exactly does each array entry of export default represent? I see a split configuration, etc. At first I thought these were something like overrides and defining linter configurations per file.

Yes, this is similar to overrides, you can define different configurations for different files:

https://eslint.org/docs/latest/use/configure/configuration-files#specifying-files-and-ignores

> ESLint does not work in VSCode anymore (no error). With CLI I see errors. But not visible in editor. Any guesses?

You can try with the following settings (`.vscode/settings.json`):

```json
{
    "eslint.experimental.useFlatConfig": true
}

```

vscode-eslint v3 prereleases support automatic detection, so this will not be needed in v3. Final v3 release is, I believe, blocked on an issue with package managers: https://github.com/microsoft/vscode-eslint/issues/1644

**Reply by fudom** [reply]
Thanks, that works. I suspected that it might not be ready yet. 

I still have some weird issues with 'ignores' and the order with the array order etc. Not sure, but I think it works different as the overrides before. I need to split the ignores in single array entry to work. When I move the ignores to the object with languageOptions or with the files, it works different. Also the order has an effect...

```js
import pluginJs from '@eslint/js';
import globals from 'globals';
import tseslint from 'typescript-eslint';

export default [
  {
    languageOptions: { globals: globals.browser },
  },
  {
    ignores: ['/*', '!/src', '**/*.spec.ts', 'src/assets/*'],
  },
  pluginJs.configs.recommended,
  ...tseslint.configs.recommended,
  {
    files: ['**/*.ts'],
    rules: {
      '@typescript-eslint/no-empty-function': 'warn',
    },
  },
];
```

Background: I want to globally ignore paths for subsequent configurations. I thought I could define a global object as in the root object of eslintrc. But when I move the 'ignores' to the object with 'languageOptions', the result is different. Similar weird behavior when I move the 'ignores' to the 'files' object below. I suspect it's because 'ignores' is affacting the related configuration in the individual object. For example, when it's only in the last obejct with 'files', the default config from eslint has no 'ignores' anymore and do the job everywhere which results in more findings. But what is the correct configuration for global ingores or config?

But I'm not done yet. Not sure if this is a bug or not. But maybe I'll know more once the configuration is complete. I just think aloud.

**Reply by mdjermanovic** [reply]
`ignores` in a config object without other keys means that those files will be will be entirely ignored:

https://eslint.org/docs/latest/use/configure/configuration-files#globally-ignoring-files-with-ignores

`ignores` in a config object with other keys means that that config object doesn't apply to those files:

https://eslint.org/docs/latest/use/configure/configuration-files#excluding-files-with-ignores

**Reply by fudom** [reply]
Ok, thanks. I'm almost done. One last thing: It seems that plugins are not working anymore. For example:

```js
import eslintPluginRegex from 'eslint-plugin-regex';

export default [
  {
    files: ['**/*.ts'],
    plugins: {
      eslintPluginRegex,
    },
    rules: {
      'regex/invalid': ['warn', [{ id: 'foo-bar', regex: 'foo-bar', message: 'Foo bar.' }]],
    },
  },
];
```

TypeError: Key "rules": Key "regex/invalid": Could not find plugin "regex".

**Reply by mdjermanovic** [reply]
`plugins` should be specified as `<plugin name>: <plugin object>`:

> ```js
> import eslintPluginRegex from 'eslint-plugin-regex';
> 
> export default [
>   {
>     files: ['**/*.ts'],
>     plugins: {
>       regex: eslintPluginRegex,
>     },
>     rules: {
>       'regex/invalid': ['warn', [{ id: 'foo-bar', regex: 'foo-bar', message: 'Foo bar.' }]],
>     },
>   },
> ];
> ```
